hydradx/model/amm/global_state.py

```diff
@@ -0,0 +1,154 @@
+from .agents import Agent
+import copy
+import random
+
+
+class AMM:
+    def __init__(self):
+        self.fail = ''
+
+    def copy(self):
+        copy_self = copy.deepcopy(self)
+        copy_self.fail = ''
+        return copy_self
+
+    def swap(
+        self,
+        old_agent: Agent,
+        tkn_sell: str,
+        tkn_buy: str,
+        buy_quantity: float = 0,
+        sell_quantity: float = 0
+    ):
+        return self.copy(), old_agent.copy()
+
+    def add_liquidity(
+        self,
+        old_agent: Agent,
+        quantity: float,
+        tkn_add: str
+    ):
+        return self.copy(), old_agent.copy()
+
+    def remove_liquidity(
+        self,
+        old_agent: Agent,
+        quantity: float,
+        tkn_remove: str
+    ):
+        return self.copy(), old_agent.copy()
+
+    def fail(self, agent: Agent, error: str = 'fail'):
+        failed_state = self.copy()
+        failed_state.fail = error
+        return failed_state, agent.copy()
+
+
+class GlobalState:
+    def __init__(self, agents: dict[str: Agent], pools: dict[str: AMM], external_market: dict[str: float]):
+        # get a list of all assets contained in any member of the state
+        self.asset_list = list(set(
+            [asset for pool in pools.values() for asset in pool.liquidity.keys()]
+            + [asset for agent in agents.values() for asset in agent.asset_list]
+            + list(external_market.keys())
+        ))
+        self.agents = agents
+        for agent_name in self.agents:
+            self.agents[agent_name].unique_id = agent_name
+        self.pools = pools
+        for pool_name in self.pools:
+            self.pools[pool_name].unique_id = pool_name
+        self.external_market = external_market
+
+    def price(self, asset):
+        return self.external_market[asset] if asset in self.external_market else 0
+
+    def copy(self):
+        self_copy = copy.deepcopy(self)
+        return self_copy
+
+
+def fluctuate_prices(state: GlobalState, percent: float, bias: float):
```

Comment:
I like that you've refactored this market evolution step out. There's very standard ways to evolve market prices under common assumptions that asset values follow brownian motion. I think this is a decent place to start for reading:
https://en.wikipedia.org/wiki/Brownian_model_of_financial_markets
I have some books on this too, it's all a little math-intensive but the actual implementations wind up being pretty intuitive. So having the ability to plug in different functions for different kinds of asset value evolution is great.